src/main/kotlin/org/rust/ide/inspections/checkMatch/Constructor.kt

```
@@ -0,0 +1,107 @@
+
+import org.rust.lang.core.psi.RsEnumItem
+import org.rust.lang.core.psi.RsEnumVariant
+import org.rust.lang.core.psi.RsStructItem
+import org.rust.lang.core.psi.ext.RsFieldsOwner
+import org.rust.lang.core.psi.ext.fieldTypes
+import org.rust.lang.core.psi.ext.size
+import org.rust.lang.core.psi.ext.variants
+import org.rust.lang.core.types.ty.*
+
+sealed class Constructor {
+
+    /** The constructor of all patterns that don't vary by constructor, e.g. struct patterns and fixed-length arrays */
+    object Single : Constructor() {
+        override fun coveredByRange(from: Constant, to: Constant, included: Boolean): Boolean = true
+    }
+
+    /** Enum variants */
+    data class Variant(val variant: RsEnumVariant, val index: Int) : Constructor()
+
+    /** Literal values */
+    data class ConstantValue(val value: Constant) : Constructor() {
+        override fun coveredByRange(from: Constant, to: Constant, included: Boolean): Boolean =
+            if (included) {
+                value >= from && value <= to
+            } else {
+                value >= from && value < to
+            }
+    }
+
+    /** Ranges of literal values (`2...5` and `2..5`) */
+    data class ConstantRange(val start: Constant, val end: Constant, val includeEnd: Boolean = false) : Constructor() {
+        override fun coveredByRange(from: Constant, to: Constant, included: Boolean): Boolean =
+            if (includeEnd) {
+                ((end < to) || (included && to == end)) && (start >= from)
+            } else {
+                ((end < to) || (!included && to == end)) && (start >= from)
+            }
+    }
+
+    /** Array patterns of length n */
+    data class Slice(val size: Int) : Constructor()
+
+    fun arity(type: Ty): Int = when (type) {
+        is TyTuple -> type.types.size
+
+        is TySlice, is TyArray -> when (this) {
+            is Constructor.Slice -> size
+            is Constructor.ConstantValue -> 0
+            else -> 0 // impossible case
```

Maybe throw an exception too?